src/libopenrave/kinbodystatesaver.cpp

```diff
@@ -18,6 +18,121 @@
 
 namespace OpenRAVE {
 
+/// \brief push link to _listNonCollidingLinksWhenGrabbed of grabbed.
+static void _PushLinkToListNonCollidingLinksWhenGrabbed(Grabbed& grabbed,
+                                                        const int linkindex,
+                                                        const KinBody::LinkConstPtr& pLinkRef,
+                                                        const std::vector<KinBody::LinkPtr>& vLinks,
+                                                        const EnvironmentBasePtr& pEnv)
+{
+    if( linkindex < 0 || linkindex >= (int)vLinks.size() ) {
+        RAVELOG_WARN_FORMAT("env=%s, could not restore link '%s' since its index %d is out of range (body num links is %d)", pEnv->GetNameId()%pLinkRef->GetName()%pLinkRef->GetIndex()%vLinks.size());
+    }
+    else {
+        grabbed._listNonCollidingLinksWhenGrabbed.push_back(vLinks.at(linkindex));
+    }
+}
+
+void KinBody::_RestoreGrabbedBodiesFromSavedData(const KinBody& savedBody,
+                                                 const int options,
+                                                 const std::unordered_map<int, GrabbedPtr>& savedGrabbedBodiesByEnvironmentIndex)
+{
+    {
+        // have to release all grabbed first
+        ReleaseAllGrabbed();
+        OPENRAVE_ASSERT_OP(_grabbedBodiesByEnvironmentIndex.size(),==,0);
+        for (const std::unordered_map<int, GrabbedPtr>::value_type& grabPair : savedGrabbedBodiesByEnvironmentIndex) {
+            const GrabbedPtr& pGrabbed = grabPair.second;
+            KinBodyPtr pGrabbedBody = pGrabbed->_pGrabbedBody.lock();
+            if( !!pGrabbedBody ) {
```

Comment:
If we invert this check (`if (!pGrabbedBody) { WARN(...); continue; }`) we can unnest a significant portion of this function

Reply:
same for the above, fixed via https://github.com/rdiankov/openrave/pull/1429/commits/4970937f95fa69faa5839465e0b3b21be80db0bf.

Thanks!